### PR TITLE
Outgoing/cleanup framework

### PR DESCRIPTION
**Pull request by scrosby**
## Changes proposed in this PR

- Cleanup how the framework is propagated throughout the code.

    
    It used to be threaded through the scheduler and api query code, stuffed
    into the job object, then threaded through fenzo, and finally sent as part
    of offer responses. Shortcut all of that indirection and source it
    directly from a global.
 

## Why are we making these changes?

It gets is closer to letting cook have multiple backends. E.g., manage multiple mesos clusters.

### CODE REVIEW

File: scheduler/src/cook/mesos/task.clj [hunk 1]
```diff
@@ -140,9 +140,9 @@
 
 (defn TaskAssignmentResult->task-metadata
   "Organizes the info Fenzo has already told us about the task we need to run"
-  [db framework-id mesos-run-as-user ^TaskAssignmentResult task-result]
+  [db mesos-run-as-user ^TaskAssignmentResult task-result]
   (let [{:keys [job task-id] :as task-request} (.getRequest task-result)]
-    (merge (job->task-metadata db framework-id mesos-run-as-user job task-id)
+    (merge (job->task-metadata db @cook.config/framework-id-atom mesos-run-as-user job task-id)
```
Comment (scrosby) [inline scheduler/src/cook/mesos/task.clj]: I kept this here. Long term when cook schedules over multiple domains, it can vary on a per-offer basis in the fenzo cycle. 

Longer term, this will be replaced with a function from the Fenzo result (indicating the scheduling domain matched) that will store the selected scheduling domain into the task structure (and subsequently use that in order to launch it in the right spot.).

Today, however, we have a single scheduling domain --- a single mesos cluster, so just stuff that in here.

we'll encode the domain in fenzo.

File: integration/tests/cook/test_basic.py [hunk 2]
```diff
@@ -565,7 +565,6 @@ def test_get_job(self):
         # 9223372036854775807 is MAX_LONG(ish), the default value for max_runtime
         self.assertEqual('success', job['state'])
         self.assertTrue('env' in job)
-        self.assertTrue('framework_id' in job)
```
[the remark this answers is not in the document]
Comment (pschorf) [reply integration/tests/cook/test_basic.py]: Are we actually removing the field? Or just removing it from the test? Either way, I'd vote to keep it. Even if it's broken, it's part of the API. We can come up with a new field but should probably keep putting something here.

File: integration/tests/cook/test_basic.py [hunk 3]
```diff
@@ -1605,7 +1606,7 @@ def get_docker_info():
             job = util.load_job(self.cook_url, job_uuid)
             self.logger.info(f'Job status is {job["status"]}: {job}')
             util.session.delete('%s/rawscheduler?job=%s' % (self.cook_url, job_uuid))
-            mesos.dump_sandbox_files(util.session, instance, job)
+            mesos.dump_sandbox_files(util.session, instance, cook_framework, cook_framework)
```
Comment (pschorf) [inline integration/tests/cook/test_basic.py]: `util.session, instance, job, cook_framework`

File: scheduler/src/cook/components.clj [hunk 4]
```diff
@@ -338,9 +337,12 @@
     (mount/start-with-args (cook.config/read-config config-file-path))
     (pool/guard-invalid-default-pool (d/db datomic/conn))
     (metrics-jvm/instrument-jvm)
-    (let [server (scheduler-server config)]
+    (let [server (scheduler-server config)
+          framework-id (get-in server [:framework-id])]
       (intern 'user 'main-graph server)
-      (log/info "Started Cook, stored variable in user/main-graph"))
+      (log/info "Started Cook, stored variable in user/main-graph")
+      (log/info "Framework-id " )
```
Comment (pschorf) [inline scheduler/src/cook/components.clj]:
```suggestion
      (log/info "Framework-id " framework-id)
```

File: scheduler/src/cook/mesos/api.clj [hunk 5]
```diff
@@ -341,8 +341,7 @@
   (-> JobRequestMap
       (dissoc (s/optional-key :group))
       (dissoc (s/optional-key :status))
-      (merge {:framework-id (s/maybe s/Str)
-              :retries-remaining NonNegInt
+      (merge {:retries-remaining NonNegInt
```
Comment (pschorf) [inline scheduler/src/cook/mesos/api.clj]: Discussed above, I don't think we should remove this, even if it's currently broken.

File: scheduler/src/cook/mesos/scheduler.clj [hunk 6]
```diff
@@ -1418,21 +1418,15 @@
...
    handle-exit-code handle-progress-message sandbox-syncer-state]
-  (let [sync-agent-sandboxes-fn #(sandbox/sync-agent-sandboxes sandbox-syncer-state configured-framework-id %1 %2)
+  (let [sync-agent-sandboxes-fn #(sandbox/sync-agent-sandboxes sandbox-syncer-state @cook.config/framework-id-atom %1 %2)
         message-handlers {:handle-exit-code handle-exit-code
                           :handle-progress-message handle-progress-message}]
     (mesos/scheduler
       (registered
         [this driver framework-id master-info]
         (log/info "Registered with mesos with framework-id " framework-id)
-        (let [value (-> framework-id mesomatic.types/pb->data :value)]
```
Comment (pschorf) [inline scheduler/src/cook/mesos/scheduler.clj]: Why was this removed?

File: scheduler/src/cook/mesos/task.clj (hunk 1 above)
Comment (pschorf) [reply scheduler/src/cook/mesos/task.clj]: Can we put a code comment reflecting what you said here?

File: integration/tests/cook/test_basic.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (scrosby) [reply integration/tests/cook/test_basic.py]: I"m reverting this part of the change. Removing the framework-id out of job is a much bigger change with more communication needed. (I'm starting on that process now.)

FWIW, I'd rather remove it than leave behind something wrong. Regardless, thats going to be on another PR, so we can discuss it then?

File: scheduler/src/cook/mesos/task.clj (hunk 1 above)
Comment (scrosby) [reply scheduler/src/cook/mesos/task.clj]: Done.

File: scheduler/src/cook/components.clj (hunk 4 above)
Comment (scrosby) [reply scheduler/src/cook/components.clj]: I wanted to remove the field from the job structure, but that change needs to be reverted as it turned out to be more invasive than expected. I also reverted the sandbox changes. No reason to change them now and again when we put it into the task structure.

File: scheduler/src/cook/mesos/scheduler.clj (hunk 6 above)
Comment (scrosby) [reply scheduler/src/cook/mesos/scheduler.clj]: It seemed unnecessary. We get the driver via code around here:

https://github.com/scrosby/Cook/blob/outgoing/cleanup-framework-id/scheduler/src/cook/mesos.clj#L197
create-datomic-scheduler calls create-mesos-scheduler.

Notice that we it what framework id we are. So if we ever get a registration callback for another framework-id, something is really screwed up; its a bug in our implementation/initialization. The driver atom is initialized about 20 lines further down.

We could retain this code and have it refer to the global variable, but I don't see the point, or the value.

File: scheduler/src/cook/components.clj (hunk 4 above)
Comment (scrosby) [reply scheduler/src/cook/components.clj]: :) Like that change.

File: scheduler/src/cook/mesos/scheduler.clj (hunk 6 above)
Comment (pschorf) [reply scheduler/src/cook/mesos/scheduler.clj]: I think the point of this is to make sure there's not something really screwed up :)

It seems pretty harmless to me as a sanity check.

Comment (scrosby) [reply scheduler/src/cook/mesos/scheduler.clj]: Left in.

**Comment by pschorf** [comment]
Build is red, otherwise LGTM

File: integration/tests/cook/test_basic.py (hunk 2 above)
Comment (pschorf) [inline integration/tests/cook/test_basic.py]: Shouldn't this be re-added?

File: scheduler/src/cook/components.clj [hunk 7]
```diff
@@ -338,9 +337,11 @@
     (mount/start-with-args (cook.config/read-config config-file-path))
     (pool/guard-invalid-default-pool (d/db datomic/conn))
     (metrics-jvm/instrument-jvm)
-    (let [server (scheduler-server config)]
+    (let [server (scheduler-server config)
+          framework-id (get-in server [:framework-id])]
```
Comment (pschorf) [inline scheduler/src/cook/components.clj]: Can this be removed?

File: scheduler/test/cook/test/mesos/api.clj [hunk 8]
```diff
@@ -1187,19 +1187,18 @@
...
                      disable-mea-culpa-retries executor datasets]
-              :or {disable-mea-culpa-retries false}}
-             framework-id]
+              :or {disable-mea-culpa-retries false}}]
             (cond-> {;; Fields we will fill in from the provided args:
                      :command command
                      :cpus cpus
                      :disable_mea_culpa_retries disable-mea-culpa-retries
-                     :framework_id framework-id
+                     :framework_id nil
```
Comment (pschorf) [inline scheduler/test/cook/test/mesos/api.clj]: Can this be removed instead of being set to `nil`?

Comment (scrosby) [reply scheduler/test/cook/test/mesos/api.clj]: No. Unit tests use this to make the pre-canned response they validate against. I either have the field, and set it to nil, or I have to wrap a bunch of unit tests with (with-redefs [framework-id-config (fn .....].

File: scheduler/src/cook/components.clj (hunk 7 above)
Comment (scrosby) [reply scheduler/src/cook/components.clj]: Good catch.

File: integration/tests/cook/test_basic.py (hunk 2 above)
Comment (scrosby) [reply integration/tests/cook/test_basic.py]: Either way.... :)

File: scheduler/test/cook/test/mesos/api.clj (hunk 8 above)
Comment (scrosby) [reply scheduler/test/cook/test/mesos/api.clj]: Did the redef as we discussed.
